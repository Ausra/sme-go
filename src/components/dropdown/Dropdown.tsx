import { FunctionComponent, useState } from "react";
import Input, { InputProps } from "../input";
import {
  DropDownListContainer,
  StyledListButton,
  StyledListItem,
} from "./styles";

export interface DropdownOption {
  value: string;
  id: string;
}

interface DropdownProps extends InputProps {
  dataTestId?: string;
  options: DropdownOption[];
  label: string;
  onSelectCallback?: (value: string) => void;
}

export const defaultTestId = "styled-dropdown";
export const dropdownListTestId = "styled-dropdown-list";
export const listItemTestId = "styled-dropdown-list-item";
export const listItemButtonId = "styled-dropdown-list-item-button";

const Dropdown: FunctionComponent<DropdownProps> = ({
  label,
  options,
  dataTestId,
  name,
  status,
  statusMessage,
  customValue,
  onSelectCallback,
}) => {
  const [open, setOpen] = useState<boolean>(false);
  const [selected, setSelected] = useState<string | undefined>(undefined);

  const handleOpen = () => {
    setOpen(!open);
  };

  const handleSelect = (value: string) => {
    setSelected(value);
    onSelectCallback && onSelectCallback(value);
    setOpen(false);
  };

  const handleOnChange = (e: any) => {
    setSelected(e.target.value);
  };

  return (
    <>
      <Input
        dataTestId={dataTestId || defaultTestId}
        name={name}
        label={label}
        onClickCallback={handleOpen}
        customValue={customValue || selected}
        onChangeCallback={handleOnChange}
        status={status}
        statusMessage={statusMessage}
      />
      {open ? (
        <DropDownListContainer data-testid={dropdownListTestId}>
          {options.map((option, index) => (
            <StyledListItem key={index} data-testid={listItemTestId}>
              <StyledListButton
                data-testid={listItemButtonId}
                onClick={() => handleSelect(option.value)}
              >
                {option.value}
              </StyledListButton>
            </StyledListItem>
          ))}
        </DropDownListContainer>
      ) : null}
    </>
  );
};

export default Dropdown;
