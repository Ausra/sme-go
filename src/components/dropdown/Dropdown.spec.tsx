import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "jest-styled-components";
import Dropdown, { defaultTestId, listItemButtonId } from "./Dropdown";

const OPTIONS = [
  { id: "1", value: "Lithuania" },
  { id: "2", value: "USA" },
];

describe("Dropdown styles", () => {
  it("renders", () => {
    render(<Dropdown label="Country of Registration" options={OPTIONS} />);

    const dropdownInput = screen.getByTestId(defaultTestId);
    expect(dropdownInput).toMatchSnapshot();
  });
});

describe("Dropdown", () => {
  it("renders", () => {
    render(<Dropdown label="Country of Registration" options={OPTIONS} />);
    const dropdownInput = screen.getByTestId(defaultTestId);
    expect(dropdownInput).toBeInTheDocument();
  });

  it("should call onSelectCallback", async () => {
    const user = userEvent.setup();
    const onSelectCallback = jest.fn();
    render(
      <Dropdown
        label="Country of Registration"
        options={OPTIONS}
        onSelectCallback={onSelectCallback}
      />
    );

    const dropdownInput = screen.getByTestId(defaultTestId);
    await user.click(dropdownInput);

    const listItem = screen.getAllByTestId(listItemButtonId);

    await user.click(listItem[0]);

    await waitFor(() => {
      expect(onSelectCallback).toHaveBeenCalledWith(OPTIONS[0].value);
    });
  });
});
