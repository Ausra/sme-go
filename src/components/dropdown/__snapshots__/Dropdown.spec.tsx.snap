// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown styles renders 1`] = `
.c0 {
  font-size: 14px;
  border: 1px solid;
  border-color: #CCCCCC;
  width: 100%;
  height: 60px;
  border-radius: 8px;
  padding-top: 16px;
  padding-left: 16px;
  background-color: #ffffff;
  outline: none;
}

.c0::-webkit-input-placeholder {
  color: transparent;
}

.c0::-moz-placeholder {
  color: transparent;
}

.c0:-ms-input-placeholder {
  color: transparent;
}

.c0::placeholder {
  color: transparent;
}

.c0:not(:placeholder-shown),
.c0:focus {
  background-color: #F7F7F7;
}

<input
  class="c0"
  data-testid="styled-dropdown"
  id="styled-input"
  placeholder=" "
  type="text"
  value=""
/>
`;
